lib/publiccloud/ssh_interactive.pm

```diff
@@ -1,67 +1,98 @@
-        cmd => "'rm -rf /dev/sshserial; mkfifo -m a=rwx /dev/sshserial; tail -fn +1 /dev/sshserial' 2>&1 | tee /dev/$serialdev; clear",
-        timeout => 0,
-        no_quote => 1,
-        ssh_opts => "-yt -R $upload_port:$upload_host:$upload_port",
-        username => 'root',
-    );
-    sleep 3;
-    save_screenshot;
+    # Pipe the output of the device fifo to the local serial terminal
+# Note: We run this in a loop so that the ssh tunnel gets automatically re-established after device reboots and such. The sleep helps to avoid unnecessary CPU hogging in case of connection issues
+    background_script_run("while true; do ssh sut -yt -R '$upload_port:$upload_host:$upload_port' 'rm -f /dev/sshserial && mkfifo -m a=rwx /dev/sshserial && tail -fn +1 /dev/sshserial' 2>&1 >/dev/$serialdev; sleep 5; done");
+    # give the ssh connection some time to settle
+    sleep 10;
 
+    # from here onwards, the serial output should be directed to /dev/sshserial instead to /dev/ttyS0
     set_var('SERIALDEV_', $serialdev);
+    set_sshserial_dev();
     set_var('_SSH_TUNNELS_INITIALIZED', 1);
-
     set_var('AUTOINST_URL_HOSTNAME', 'localhost');
-    set_sshserial_dev();
+
+    # When re-activing the consoles we also need to setup the underlying ssh connections again,
+    # because this only happens the first time the console is activated
+    if ($args{reconnect}) {
+        establish_tunnel_console("root-console");
+        establish_tunnel_console("user-console");
+        establish_tunnel_console("root-virtio-terminal");
+    }
+
+    select_console($prev_console) if ($prev_console !~ /tunnel-console/);
 }
 
 sub ssh_interactive_leave {
-    # Check if the SSH tunnel is still up and leave the SSH interactive session
-    script_run("test -p /dev/sshserial && exit", timeout => 0);
-
-    # Restore the environment to not use the SSH tunnel for upload/download from the worker
-    #set_var('SUT_HOSTNAME',          testapi::host_ip());
-    set_var('AUTOINST_URL_HOSTNAME', testapi::host_ip());
+    my $prev_console = current_console();
+    # Switch to the local console, terminte the ssh tunnel and set the serial device back to ttyS0
+    select_console('tunnel-console') unless ($prev_console =~ /tunnel-console/);
     unset_sshserial_dev();
+    set_var('AUTOINST_URL_HOSTNAME', testapi::host_ip());
 
-    $testapi::distri->set_standard_prompt('root');
+    # Terminate the ssh tunnel.
+    # It can happen that the serial console drops some characters thereby we might need to try this multiple times
+    my $retries = 8;
+    while ($retries-- > 0) {
+        type_string('kill %1', max_interval => 1, lf => 1);    # Slow typing reduces the chance of characters being dropped
+        last if (script_run('true', die_on_timeout => 0) == 0);
+    }
+    die "tunnel-console is not terminated" if ($retries <= 0);
+    assert_script_run("true", fail_message => "tunnel-console is not functional");    # additional health check
```

Comment:
i dont understand why do you need this after $retries. So `while` can return true but still fail in the next attempt?

Reply:
Not strictly necessary, if should already work with the ones before.

Still because there is a (low but still) risk that the console is not functional at this point, the check is required as a last barrier to really ensure, everything works as expected.

Reply:
I changed the order, the first true is actually not needed. Please recheck.